### DISCUSSION TITLE
QUESTION: client.write_gatt_char(sUUID, info) data not received by server / no response

### DISCUSSION THREAD

**Discussion by WouterJD** [hbldh/bleak#772] [Q&A] [answered]
* bleak version: current
* Python version: 3
* Operating System: Windows 10

### Description

I am writing a Fitness Machine Server and to be able to debug the server, also a Collector (client).

The collector is based upon bleak\examples\discover.py, service_explorer.py

### What is already working

As a service an existing FTMS is used and that service is correctly detected.

### What is NOT yet working

When sending data to the server "await client.write_gatt_char(sUUID, info)" it seems the data does not arrive and no response is received.
The sUUID is the "Fitness Machine Control Point" props=['write', 'indicate']

Should such a UUID also be linked as "await client.start_notify(uuid, notificationHandler)"?

**In other words: what is the calling sequence to write a characteristic of a server?**

**Comment by dlech** [comment]
If you expect a response via an indication, then yes, you need to register a callback with `start_notify()`.

If you still have problems, see the troubleshooting section in the docs.

**Comment by WouterJD** [comment]
Thanks, I will check the docs (again, did not yet find the correct spot).
And yes, I registered with start_notify() , but it's not notified.

I'll be back😉 Thanks again for this library, I'm at 98% of what I need. Last step always the hardest.

**Comment by WouterJD** [comment]
Hi David, perhaps you can help me.
You analysis would be highly appreciated.

Based upon your examples, I have written a Collector, being a client for a Fitness Machine. [Source module is here](https://raw.githubusercontent.com/WouterJD/FortiusANT/Implement-BLE-using-bless/pythoncode/bleCollector.py)
This collector prints the structure and is inspired by discover.py, service_explorer.py and enable_notifications.py; this works.

start_notify() also works, since notifications are received; [At this line](https://github.com/WouterJD/FortiusANT/blob/Implement-BLE-using-bless/pythoncode/bleCollector.py#L204)

From then it's unclear [At this line](https://github.com/WouterJD/FortiusANT/blob/Implement-BLE-using-bless/pythoncode/bleCollector.py#L234)

After `await client.write_gatt_char(bc.cFitnessMachineControlPointUUID, info)` I would expect an indication, but nothing is received.

---

The printed output is as below; no response on the FitnessMachineControlPoint is received; other notifications are.
There is no logging output in these modules, so I have no feedback from bleak internal


```
Register notifications
Register indications
Request control, so that commands can be sent
30 Heart Rate Measurement "00 00"                   speed=0 cadence=  0 power=   0 hrm=  0
14 Indoor Bike Data       "44 00 00 00 c6 00 67 00" speed=0 cadence= 99 power= 103 hrm=  0
14 Indoor Bike Data       "44 00 00 00 c8 00 63 00" speed=0 cadence=100 power=  99 hrm=  0
30 Heart Rate Measurement "00 00"                   speed=0 cadence=100 power=  99 hrm=  0
Request control; waiting for response
14 Indoor Bike Data       "44 00 00 00 c6 00 61 00" speed=0 cadence= 99 power=  97 hrm=  0
30 Heart Rate Measurement "00 00"                   speed=0 cadence= 99 power=  97 hrm=  0
Request control; waiting for response
14 Indoor Bike Data       "44 00 00 00 ca 00 66 00" speed=0 cadence=101 power= 102 hrm=  0
30 Heart Rate Measurement "00 00"                   speed=0 cadence=101 power= 102 hrm=  0
14 Indoor Bike Data       "44 00 00 00 c6 00 62 00" speed=0 cadence= 99 power=  98 hrm=  0
30 Heart Rate Measurement "00 00"                   speed=0 cadence= 99 power=  98 hrm=  0
Request control; waiting for response


```

**Comment by WouterJD** [comment]
Note; all the information above is on Windows 10 with a BLE 5.0 dongle.

The output below is on a raspberry Pi0 W:

```
Register notifications
14 Indoor Bike Data       "44 00 00 00 c4 00 65 00" speed=0 cadence= 98 power= 101 hrm=  0
30 Heart Rate Measurement "00 56"                   speed=0 cadence= 98 power= 101 hrm=22016
Register indications
Request control, so that commands can be sent
indicationFitnessMachineControlPoint() TO BE IMPLEMENTED <--------
22 Fitness Machine Control Point "80 00 05"                speed=0 cadence= 98 power= 101 hrm=22016
14 Indoor Bike Data       "44 00 00 00 cc 00 61 00" speed=0 cadence=102 power=  97 hrm=22016
```
**This means that the indication is received on RPI0**
`22 Fitness Machine Control Point "80 00 05"                speed=0 cadence= 98 power= 101 hrm=22016`


**So this shows that there is a difference between the two implementations**

**Comment by dlech** [comment]
Logging Bluetooth packets will let you see what is actually being sent to the Bluetooth chip. If you do this on both Windows and Linux, you should be able to spot the difference. In the winrt backend, there is a `force_indicate` kwarg to `start_notify()` to force enabling indications. However, this should not be needed in this case if the characteristic only has the indicate property and not the notify property as you have stated.

**Reply by WouterJD** [reply]
> Logging Bluetooth packets will let you see what is actually being sent to the Bluetooth chip.

How do I activate that? Please.

**Reply by dlech** [reply]
It is described in the troubleshooting section of the docs.

**Comment by WouterJD** [comment]
Hi, thanks for staying with me.
I'm now testing on RPI0 which shows good results as expected.

As soon as it works, I will move back to testing on Windows 10 to check whether the indications actually do not appear.
Moving forward :-)

**Comment by WouterJD** [comment]
https://github.com/WouterJD/FortiusANT/blob/Implement-BLE-using-bless/pythoncode/bleClient.py
is now available.
This program can be used to validate and FTMS (Fitness Machine Service).

The program works on Raspberry rpi0W raspbian (10) buster
It does not work well on Windows 10 (indications not received), reason unknown.

On Windows 10, Trainer Road can work succesfully with the FTMS server; so I assume BLEdongle and system-drivers are ok.

**Comment by WouterJD** [comment, marked as answer]
The way how to work has become clear now
indications are received on Raspberry Pi0 but do NOT work on Windows 10; new issue written #775
